pkg/frontend/querymiddleware/codec_test.go

```diff
@@ -99,14 +133,310 @@ func TestMetricsQueryRequest(t *testing.T) {
+			expectedMinT: startTime.UnixMilli() - offsetDurationMS,
+			expectedMaxT: endTime.UnixMilli() - offsetDurationMS,
+		},
+		{
+			name:         "instant query: without range vector, with offset",
+			metricsQuery: withQuery(t, instantRequest, fmt.Sprintf("go_goroutines{} offset %s", offsetDurationStr)),
+			expectedMinT: endTime.UnixMilli() - offsetDurationMS,
+			expectedMaxT: endTime.UnixMilli() - offsetDurationMS,
+		},
+		{
+			name:         "range query: with range vector, with offset",
+			metricsQuery: withQuery(t, rangeRequest, fmt.Sprintf("rate(go_goroutines{}[%s] offset %s)", rangeVectorDurationStr, offsetDurationStr)),
+			expectedMinT: startTime.UnixMilli() - rangeVectorDurationMS - offsetDurationMS,
+			expectedMaxT: endTime.UnixMilli() - offsetDurationMS,
+		},
+		{
+			name:         "instant query: with range vector, with offset",
+			metricsQuery: withQuery(t, instantRequest, fmt.Sprintf("rate(go_goroutines{}[%s] offset %s)", rangeVectorDurationStr, offsetDurationStr)),
+			expectedMinT: endTime.UnixMilli() - rangeVectorDurationMS - offsetDurationMS,
+			expectedMaxT: endTime.UnixMilli() - offsetDurationMS,
+		},
+		// permutations with and without range vectors and @ modifiers
+		{
+			name:         "range query: with @ modifer",
+			metricsQuery: withQuery(t, rangeRequest, fmt.Sprintf("go_goroutines{} @ %d", endTime.Add(-atModifierDuration).Unix())),
+			expectedMinT: endTime.Add(-atModifierDuration).UnixMilli(),
+			expectedMaxT: endTime.Add(-atModifierDuration).UnixMilli(),
+		},
+		{
+			name:         "instant query: with @ modifer",
+			metricsQuery: withQuery(t, instantRequest, fmt.Sprintf("go_goroutines{} @ %d", endTime.Add(-atModifierDuration).Unix())),
+			expectedMinT: endTime.Add(-atModifierDuration).UnixMilli(),
+			expectedMaxT: endTime.Add(-atModifierDuration).UnixMilli(),
+		},
+		{
+			name:         "range query: with range vector, with @ modifer",
+			metricsQuery: withQuery(t, rangeRequest, fmt.Sprintf("go_goroutines{}[%s] @ %d", rangeVectorDurationStr, endTime.Add(-atModifierDuration).Unix())),
+			expectedMinT: endTime.Add(-(atModifierDuration + rangeVectorDuration)).UnixMilli(),
+			expectedMaxT: endTime.Add(-atModifierDuration).UnixMilli(),
+		},
+		{
+			name:         "instant query: with range vector, with @ modifer",
+			metricsQuery: withQuery(t, instantRequest, fmt.Sprintf("go_goroutines{}[%s] @ %d", rangeVectorDurationStr, endTime.Add(-atModifierDuration).Unix())),
+			expectedMinT: endTime.Add(-(atModifierDuration + rangeVectorDuration)).UnixMilli(),
+			expectedMaxT: endTime.Add(-atModifierDuration).UnixMilli(),
+		},
+	} {
+		t.Run(testCase.name, func(t *testing.T) {
+			minT := testCase.metricsQuery.GetMinT()
+
+			maxT := testCase.metricsQuery.GetMaxT()
+
+			require.EqualValues(t, testCase.expectedMinT, minT)
+			require.EqualValues(t, testCase.expectedMaxT, maxT)
+		})
+	}
+}
+
+func TestMetricsQuery_MinMaxTime_TransformConsistency(t *testing.T) {
```

Comment:
Have you considered splitting this into two parts, one to test moving the start and end times and the other to test changing the expression? This test never tests both at once, and it would remove a bunch of the conditional logic here, making the test intent clearer.